[Website Content] Add pattern documentation  & overview (#815)
hashicorp/design-system: website/docs/patterns/overview.md

---
title: Patterns
status: comingSoon
---

## What is a pattern?

A pattern is a combination of components that work together to accomplish a function or render a layout. They are often used to solve the same problem consistently. Common examples of patterns are forms, navigation, button sets, etc.

## HDS pattern support

> Patterns are not currently supported by the HashiCorp Design System, but there are future plans to introduce guidelines, documentation, and best practices around assembling HDS components in a consistent and scalable manner.

Reviewer: ```suggestion
We don't currently support patterns, but we have a few on our [roadmap](URL). We're also planning to introduce guidelines and best practices around assembling components in a consistent and scalable manner.
```

Patterns are not currently supported by the HashiCorp Design System, but there are future plans to introduce guidelines, documentation, and best practices around assembling HDS components in a consistent and scalable manner.

!!! Info

> **Note:** There are a small number of HDS components that showcase a pattern-like level of complexity ([Modal](/components/modal), [Table](/components/table), [Pagination](/components/pagination)) but are considered components and published as such.

Reviewer 2: ```suggestion
**Note:** 

There are a small number of HDS components that showcase a pattern-like level of complexity ([Modal](/components/modal), [Table](/components/table), [Pagination](/components/pagination)) but are considered components and published as such.
```

---
title: Patterns
status: comingSoon
---

## What is a pattern?

> Pattern generally refers to a combination of components which work together to accomplish a function or render a layout and is often used to solve the same problem in a consistent manner. Common examples of patterns are forms, navigation, combinations of action buttons, etc.

Reviewer: ```suggestion
A pattern is a combination of components that work together to accomplish a function or render a layout. They are often used to solve the same problem consistently. Common examples of patterns are forms, navigation, button sets, etc.
```

## HDS pattern support

> We don't currently support patterns, but we have a few on our [roadmap](URL). We're also planning to introduce guidelines and best practices around assembling components in a consistent and scalable manner.

Reviewer 3: use https://go.hashi.co/hds-rollout

> > "Components are an interface chunk to be added to an overall layout. whereas patterns may be UI or a variety of other things, like a behavior, flow, application motif or something else." ([Nathan Curtis](https://medium.com/eightshapes-llc/patterns-components-2ce778cbe4e8))

Reviewer: This doesn't read very clearly, and your description provides enough essential detail. If the overall article feels useful, maybe link to it at the bottom of the page as an additional resource.

Author: I debated leaving that in there, happy to remove it though if my description is detailed enough!

!!! Info

**Note:**

There are a small number of HDS components that showcase a pattern-like level of complexity ([Modal](/components/modal), [Table](/components/table), [Pagination](/components/pagination)) but are considered components and published as such.
!!!

If you find yourself building the same pattern often, [submit a new pattern request](https://docs.google.com/forms/d/e/1FAIpQLScpMXgrUTVT5fYriu4Pp48r4Nl_eCPluVnJLg0Yg3NXsRWvIA/viewform). 

> Don't hesitate to contact the design system team for best practices and suggestions when building and defining local components and patterns.

Reviewer 2: @jorytindall if you want to link to, now there is a `support` page/route you can use: https://hds-website-hashicorp.vercel.app/support/
just use this `<LinkTo @route="support">...</LinkTo>` (`<LinkTo>` is like `<a>` but for links within the app, so it doesn't trigger a page reload)

> Don't hesitate to contact the design system team for best practices and suggestions when building and defining local components and patterns.

Reviewer: tbh, it feels odd to send them to the support page where they'll have to click another link to get to the right form, when we can directly link to that form here.

> Don't hesitate to contact the design system team for best practices and suggestions when building and defining local components and patterns.

Author: I don't have a strong opinion here, but we should establish some consistency going forward. For these types of resources we:

* Option A: link directly to the resource from within the website
* Option B: link to the support page and from the support page link to various resources

I would advocate for option B: if we link to the support page throughout the site, and the issue form link changes or we use a different method it only has to be changed in one place.

> Don't hesitate to contact the design system team for best practices and suggestions when building and defining local components and patterns.

Reviewer: I'd advocate for Option A: if we know the type of support they need, we should send them there directly vs making them have to figure out which of the support links is the correct one when they get to the support page. I get the maintenance issue/concern though, so maybe we aim to use the same language for the link (eg. "submit a new request") so we can use a codemod to update it in the future?

If going the support link route, can we use different language for the link? Something more relevant to being taken to a general support page.

> Don't hesitate to contact the design system team for best practices and suggestions when building and defining local components and patterns.

Author: Given our conversation this morning, I'll update this to go directly to the google doc. Then we can do a bulk update for everything once we figure out a strategy for using the link shortener. Does that sound good for now?

> !!!

Reviewer: What value is the consumer getting from knowing this information?

Author: I'm trying to express here that we have these components (modal, table, etc) that might have the same characteristics as a pattern (which they do), but we don't treat them such from a technical perspective. If you feel like this is a bit irrelevant, happy to kill it.

Reviewer: It doesn't feel relevant to me, but I'm curious about what others think.

Reviewer 2: @jorytindall I agree with @heatherlarsen (also, is the most predominant thing in the page, and draws attention to it, so it's the first thing one will read - I did! - which I don't think is the intention)

Author: I'll remove it

> For up-to-date status on pattern, component, and foundation work, visit our [roadmap](/updates/roadmap).

Reviewer 3: this should link to our jira rollout tracker. we will not have this on the website at launch.

> For up-to-date status on pattern, component, and foundation work, visit our [roadmap](/updates/roadmap).

Reviewer 3: https://go.hashi.co/hds-rollout

> If you find yourself building the same type of pattern frequently using HDS components, consider [contributing](/getting-started-contribution) back to the HashiCorp Design System and making your work available to other designers and developers!

Reviewer: We don't accept pattern contributions at this time. This is great for the future, but we should keep a consistent message about what contributions are accepted. Maybe instead, the call to action is to submit a request.

Author: Ok, I'll pivot the language to point towards submitting a request

Also thinking on it, does it make sense to include something around reaching out to us for suggestions and best-practices when building local components (referencing our earlier thread about local components within the flyout)?

Reviewer: Ohh, good question. I know DesignOps is working on securing some educational resources that may help, but it couldn't hurt to let folks know we're happy to help. 🙂

> For more up-to-date status on pattern, component, and foundation work visit our [roadmap](#).

Reviewer: Love this 🎉 

[nit]
```suggestion
For up-to-date status on pattern, component, and foundation work, visit our [roadmap](#).
```
